File: 1117-as-far-from-land-as-possible/as-far-from-land-as-possible.py
from collections import deque
class Solution:
    def maxDistance(self, grid: List[List[int]]) -> int:
        n=len(grid)
        queue: deque[List[int]] = deque()
        visited:List[List[bool]]=[[False for _ in range(n)] for _ in range(n)]
        for i in range(0,n):
            for j in range(0,n):
                if grid[i][j]==1:
                    queue.append([i,j,0])
                    visited[i][j]=True
        drow=[0,0,-1,1]
        dcol=[1,-1,0,0]
        ans=-1
        while len(queue) is not 0:
            row,col,dist=queue.popleft()
            for i in range(4):
                nrow=row+drow[i]
                ncol=col+dcol[i]
                if nrow>=0 and nrow<n and ncol>=0 and ncol<n and not visited[nrow][ncol]:
                    ans=max(ans,dist+1)
                    queue.append([nrow,ncol,dist+1])
                    visited[nrow][ncol]=True
        return ans            

            

File: 1117-as-far-from-land-as-possible/as-far-from-land-as-possible.java
class Solution {
    public int maxDistance(int[][] grid) {
        int n=grid.length;
        boolean[][]visited=new boolean[n][n];    
        Queue<int[]>queue=new LinkedList<>();
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j]==1){
                    queue.add(new int[]{i,j,0});
                    visited[i][j]=true;
                }
            }
        }
        int[]drow={0,0,-1,1};
        int[]dcol={1,-1,0,0};
        int ans=-1;
        while(!queue.isEmpty()){
            int[]curr=queue.poll();
            int row=curr[0],col=curr[1],dist=curr[2];
            for(int i=0;i<4;i++){
                int nrow=row+drow[i];
                int ncol=col+dcol[i];
                if(nrow>=0 && nrow<n && ncol>=0 && ncol<n && !visited[nrow][ncol]){
                    ans=Math.max(ans,dist+1);
                    queue.add(new int[]{nrow,ncol,dist+1});
                    visited[nrow][ncol]=true;
                }
            }
        }
        return ans;

    }
}